### PR TITLE
[github-actions] use `concurrency` feature to cancel outstanding jobs

### PR DESCRIPTION
**Pull request by jwhui**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -30,15 +30,11 @@ name: Build
 
 on: [push, pull_request]
 
-jobs:
+concurrency:
+  group: ${{ github.workflow }}-${{ (github.ref == 'refx/heads/main' && github.run_id) || github.event.pull_request.number || github.ref }}
```
Comment (simonlingoogle) [inline .github/workflows/build.yml:34]:
```suggestion
  group: ${{ github.workflow }}-${{ (github.ref == 'refs/heads/main' && github.run_id) || github.event.pull_request.number || github.ref }}
```

Is it `refs` or `refx` ?

**Comment by bukepo** [comment]
let me try if i can verify this on main branch.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -30,15 +30,11 @@ name: Build
 
 on: [push, pull_request]
 
-jobs:
+concurrency:
+  group: ${{ github.workflow }}-${{ (github.ref == 'refs/heads/main' && github.run_id) || github.event.pull_request.number || github.ref }}
```
Comment (bukepo) [inline .github/workflows/build.yml:34]: Is it possible one day we have some release branches and also want this behavior of main branch on those branches?
```suggestion
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.run_id }}
```

**Review by bukepo — APPROVED** [review]

**Comment by bukepo** [comment]
> let me try if i can verify this on main branch.

Verified on my fork it works as expected:
* runs will not be cancelled by new push on main branch
* runs will be cancelled by new push on non-main branch
* runs will be cancelled by new push for pull requests

File: .github/workflows/build.yml (hunk 2 above)
Comment (bukepo) [reply .github/workflows/build.yml:34]: I think we can deal with it once we have release branches in future. resolve it now.

Comment (simonlingoogle) [reply .github/workflows/build.yml:34]: Different branches should not cancel each other.
I think github.run_id already works as intended.

What issue have we not deal with?

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1:

File: .github/workflows/build.yml (hunk 2 above)
Comment (bukepo) [reply .github/workflows/build.yml:34]: say we created a release branch `stable`, we may want runs on the `stable` branch won't be cancelled just like `main`.

Comment (simonlingoogle) [reply .github/workflows/build.yml:34]: I think `${{ github.workflow }}-${{ github.event.pull_request.number || github.run_id }}` is simpler and works better, too.

Comment (jwhui) [reply .github/workflows/build.yml:34]: I'd also like to cancel previously running jobs when pushing to a development branch in my fork.

Comment (bukepo) [reply .github/workflows/build.yml:34]: I'm fine with the current rule. And maybe we can add additional `github.repository` check.
```suggestion
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || (github.repository == 'openthread/openthread' && github.run_id) || github.ref }}
```

**Pull request merged**